MeesGame/AI/FloodFill.cs

```diff
@@ -0,0 +1,176 @@
+                { KeyColor.KeyCyan, 2},
+                { KeyColor.KeyGreen, 4},
+                { KeyColor.KeyMagenta, 8},
+                { KeyColor.KeyRed, 16},
+                { KeyColor.KeyYellow, 32}
+            };
+        
+        IAIPlayer player;
+        private Stack<Point> currentRoute = new Stack<Point>();
+        private Stack<Point> bestRoute = new Stack<Point>();
+        private bool[,] visited;
+        private int pickedUpKeys;
+        private List<PlayerAction> path = new List<PlayerAction>();
+        private TileField tileField;
+        
+        public void GameStart(IAIPlayer player, int difficulty)
+        {
+            this.player = player;
+
+            // Assign the tileField, and create the visited array with the right dimensions.
+            tileField = this.player.DummyPlayer.TileField;
+            visited = new bool[tileField.Columns, tileField.Rows];
+
+            // Start the FloodFill algorithm.
+            Flood(player.DummyPlayer.Location);
+
+            // Turn the route into a list of actions for UpdateNextAction().
+            RouteToPath(bestRoute);
+        }
+
+        public void ThinkAboutNextAction()
+        {
+        }
+
+        public void UpdateNextAction()
+        {
+            if (path.Count > 0)
+            {
+                player.NextAIAction = path[path.Count - 1];
+                path.RemoveAt(path.Count - 1);
+            }
+            else
+            {
+                player.NextAIAction = PlayerAction.NONE;
+            }
+        }
+
+        /// <summary>
+        /// The main method of the FloodFill AI.
+        /// </summary>
+        /// <param name="p"> Point to start/continue the flooding from. </param>
+        private void Flood(Point p)
+        {
+            // If we have already visited this tile, return.
+            if (visited[p.X, p.Y])
+                return;
+
+            // The current tile has been visited.
+            visited[p.X, p.Y] = true;
+
+            // GuardTile is subclass of HoleTile so this will return for both.
+            // Can't use the same for walls because DoorTile is a subclass of WallTile.
+            if (tileField.GetTile(p.X, p.Y) is HoleTile || tileField.GetType(p.X, p.Y) == TileType.Wall)
+                return;
+
+            // Add the current tile to the route.
+            currentRoute.Push(p);
+
+            // Check if the current tile is a keytile,
+            // and whether we have already picked up a key of this color.
+            if (tileField.GetTile(p.X, p.Y).TileType == TileType.Key &&
+                (pickedUpKeys & keys[(tileField.GetTile(p.X, p.Y) as KeyTile).KeyColor]) == 0)
+            {
+                // Add the key to the picked up keys.
+                pickedUpKeys |= keys[(tileField.GetTile(p.X, p.Y) as KeyTile).KeyColor];
+
+                // We need to flood everything again because there could be doors that are open now.
+                UnvisitEverything();
```

Comment:
It only erases everything if a NEW key has been picked up, that's when you would switch to a new 3d dimension, which is filled with 'false', and you will never go back to the one that is partly filled, so I don't know what the advantage of making a separate array for all the values of the keys would be.